doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,216 @@
+   user to manually compensate for the time delta between the timer has
+   expired and the callback is actually executed
+
+# 2. Requirements
+
+The following lists of requirements consist of qualitative and quantitative
+requirements. For the quantitative requirements, any improvement will enable
+more scenarios to use the timer API. Thus, no meaningful lower bounds can be
+specified. Instead, a set of benchmarks will be developed that allow to
+continuously improve the performance and to detect any possible regressions of
+future changes.
+
+## Feature Requirements
+
+- Software Timers
+    - Upper bound for number of timers only given by available resources
+    - APIs to set and abort timers
+    - Support for periodic timers (support via explicit calls to some
+      `timer_reset()` like in Contiki is fine)
+    - Long term low resolution software timers that play well with power management
+    - Short term high resolution software timers
+- Delay of program execution
+    - API to block calling thread for long periods and allows other threads to run
+    - Long term delays must play well with power management
+    - Highly precise short term delays (e.g. for bit banging)
+- System time
+    - APIs to get the current system time in reasonable units
+- Utilities for convenience
+    - Timer utilities
+        - Timer that send a message upon expiration
+        - Timer that post an event upon expiration (to allow execution in thread context)
+        - Timer that wakes up a thread upon expiration
+    - Timeout utilities
+        - Timeout for receiving a message
+        - Timeout for locking a mutex
+
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time. More
+  specifically: With *n* software timers used in parallel, the longest period of
+  time with interrupts disabled be at most *O(n)*
+- The implementation must aim for a minimized number of IRQs caused
+- The implementation must aim for keeping its ISRs short; with *n* software
+  timers used in parallel, the ISR should run for at most *O(n)*
+- The implementation should provide means to execute callbacks in thread
+  context
+
+## Precision Requirements
+
+- Timer callbacks should never fire prematurely (±1 tick of the underlying
+  timer)
+- The accuracy of execution delay should be measured with single digit micro
+  seconds for short delays in absence of other load on the system
+    - It is acceptable if platform specific tweaking is needed to achieve this
+    - It is acceptable if minimum hardware requirements are needed to achieve
+      this
+- Timer callbacks should fire closely after the timer expired
+    - In absence of other load on the system, accuracy in the order of single
+      digit micro seconds should be achievable when given hardware requirements
+      are fulfilled and platform specific tweaking is done
+
+## Performance Requirements
+
+- Efficient use of RAM/ROM
```

Comment:
@maribu @kaspar030 Here, and wherever else applicable in this section, what about giving *some* ballpark figures that we aim for, at the minimum?
For that, a (low) hanging fruit could be to use the output of some xtimer benchmarks.
In my mind, such figures would not aim to be "exhaustive", but rather to be indicative.
I think it could be helpful. What do you think?

Reply:
#12942 might provide some, but they're very dependent on both platform and configuration...

Reply:
> very dependent on both platform and configuration...

yup. That's why I was suggesting we do not claim these numbers capture *exhaustively* the requirements, but rather are put here as *indicators*, to give a rough idea?